### ISSUE TITLE
tf.select unnecessarily propagates NaNs in gradients

### ISSUE THREAD

**Issue by TimSalimans** [tensorflow/tensorflow#2540]
tf.select() will give NaN gradients even if the source of the NaNs is not selected. See example script below.

import tensorflow as tf
x = tf.placeholder(tf.float32)
y = tf.select(x>0, 0., tf.exp(x))
z = tf.gradients(y,x)[0]

with tf.Session() as sess:
    yv,zv = sess.run([y,z],{x: 1e10})
    print(yv) # correctly outputs 0
    print(zv) # this is NaN, but should be 0

**Comment by yaroslavvb** [comment]
Possibly related, Select gradient is also failing the test when there's GPU

bazel test -j 1 -c opt --config=cuda tensorflow/core:ops_math_grad_test --test_output=streamed

```
[ RUN      ] MathGradTest.Select
I tensorflow/core/common_runtime/gpu/gpu_device.cc:782] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 980, pci bus id: 0000:03:00.0)
I tensorflow/core/common_runtime/gpu/gpu_device.cc:782] Creating TensorFlow device (/gpu:1) -> (device: 1, name: GeForce GTX 980, pci bus id: 0000:04:00.0)
E tensorflow/core/common_runtime/executor.cc:334] Executor failed to create kernel. Not found: No registered '_Arg' OpKernel for GPU devices compatible with node n0 = _Arg[T=DT_BOOL, index=0]()

```

**Comment by girving** [comment]
This propagation is inevitable, actually.  If we introduce the extra variable

    w = tf.exp(x)

then `tf.gradients` correctly computes `dy/dw = 0`.  It then computes

    dy/dx = w * dy/dw = nan * 0 = nan

I.e., the issue is unrelated to select.

@yaroslavvb: That seems like an unrelated issue, so it should be filed separately if it's still a problem.

**Comment by trinity77** [comment]
Trigger update_date

**Comment by lightcatcher** [comment]
@girving It's not clear to me why the propagation of NaN is inevitable, and I think this issue should be re-opened and fixed.

If ``w = tf.select(c, t, e)``, then ``dw/dx = select(c, dt/dx, de/dx)`` 

Tensorflow 0.9 (and likely 0.10, but have not tested) do not compute this correctly. For instance,
``d/dx[ select(True, t, e) ] != dt/dx`` for some cases.

See the following example:
```python
import tensorflow as tf

x = tf.constant(0.0)

t = x
e = 1. / x

w = tf.select(True, t, e)

with tf.Session() as sess:
    print 'w:', sess.run(w)
    print 'dw/dz:', sess.run(tf.gradients(w, x)[0])
    print 'dt/dz:', sess.run(tf.gradients(t, x)[0])
```

outputs (with Tensorflow 0.9)
```
w: 0.0
dw/dz: nan
dt/dz: 1.0
```

These NaN's cause NaN gradients, which causes NaN weights and unnecessarily breaks training a net.

**Comment by lightcatcher** [comment]
After reading some code and docs, I think I see the difficulty with ``select`` gradient.

The ``select`` gradient is here: https://github.com/tensorflow/tensorflow/blob/4addf4b5806cd731949c6582a83f5824599cd1ef/tensorflow/python/ops/math_grad.py#L688

Gradients are only taken on the output of an op with respect to the input (and then multiplied by the accumulated gradient from objective to output of op), so in my example above the ``select`` gradient computes ``dw/dt=1`` and ``dw/de=0``. This gives ``dL/dt=dL/dw`` and ``dL/de=0`` for objective ``L``.

However, the graph traversal algorithm to compute gradient continues, and computes ``dt/dx = 1, de/dx = -Infinity``, which finally gives ``dL/dx = dL/dt * dt/dx + dL/de * de/dx = dL/dt * 1 + 0 * Inf = dL/dt + NaN = NaN``.

The issue is the ``0 * Inf = NaN`` term, and the fact that this computation doesn't happen in the select node (happens in the divide node in this case).

I spent a few minutes thinking of possible solutions to this, and none of them are pretty or seem worth implementing. 
One option is to pass a mask of hard zero gradients around during gradient computation. In this case, after each node computes it's gradients, the hard zero gradients can be reset to 0. This isn't a very appealing option.
The other option is to significantly change the gradient propagation algorithm. Rather than computing gradient of objective with respect to op input, compute gradient of op output with respect to op input for all ops first. Then do a multiplicative accumulate pass. This has the downside of significant framework rewrite, and is unworkable because gradient of op output wrt op input can be high order tensors (gradient of matrix multiply wrt inputs is very large).
In summary, there doesn't seem to be a good way to make Tensorflow compute the select gradient correctly. I wrote this largely for my own understanding of why "[NaN gradient] is inevitable", hopefully this can be useful to someone else who happens across the issue.

**Comment by yaroslavvb** [comment]
BTW, it's not just the issue with select, code below also gives NaN.

x = tf.placeholder(tf.float32)
y = tf.exp(x)
z = tf.exp(-y)
grad = tf.gradients(z,[x])[0]
print sess.run(grad, feed_dict={x: 1e10})

The underlying cause is that  chain rule gives f' * g' as derivative of
f(g(x)) so with because 1e10 overflows we get infinity*0 in both cases.
One could argue that limiting arguments that justify inf*0=NaN should not
apply in case of tf.select, and substitute special gradient for tf.select
using gradient_override_map during runtime

**Comment by lightcatcher** [comment]
The exponent example is a little different because a intermediate computation (``y`` in your example) is infinity, so it's not too surprising that the gradient is also problematic. The ``epsilon * Inf`` is fundamental to the gradient computation here, while it's not obviously so in the ``select`` case.

I completely agree that limiting arguments for ``0 * Inf = NaN`` do not apply for select, because the gradient in select is set to 0 to specify that the gradient with respect to the entire non-selected subgraph is exactly 0.
I wasn't familiar with ``gradient_override_map``, but I just read the documentation and I don't see how it can fix the problem as I believe the problem comes from the combination of backprop (multiplicatively accumulating gradients) and that ``0 * x != 0`` for all ``x``. @yaroslavvb , did you have a specific idea about how to fix ``select`` with ``gradient_override_map``?

**Comment by akosiorek** [comment]
I also had this problem and I needed to use a workaround. Instead of writing `log = tf.where(non_zero, tf.log(logarg), tf.zeros_like(logarg))` I needed to select values with `tf.boolean_mask` and then scatter them back into the original shape by doing:
```
logarg = tf.boolean_mask(logarg, non_zero)
log = tf.log(logarg)
idx = tf.to_int32(tf.where(non_zero))
log = tf.scatter_nd(idx, log, tf.shape(non_zero))
```

**Comment by anishathalye** [comment]
Here's another way to work around this issue (implementing what Tim was doing in the original post):

```python
import tensorflow as tf

x = tf.placeholder(tf.float32)
# y = tf.where(x > 0, 0., tf.exp(x))

# trick: we're not using the result of safe_exp when x > 0, so we can
# substitute a safe value for x in that case
# it doesn't really matter what we put in here, as long as the backward pass
# returns some finite value
safe_exp = tf.exp(tf.where(x > 0, 1.0, x))
y = tf.where(x > 0, 0., safe_exp)

z, = tf.gradients(y, x)

sess = tf.InteractiveSession()
yv, zv = sess.run([y, z], {x: 1e10})
print(yv) # correctly outputs 0
print(zv) # now correctly outputs 0

yv, zv = sess.run([y, z], {x: -1})
print(yv) # correctly outputs e^(-1)
print(zv) # correctly outputs e^(-1)
```

**Comment by jvdillon** [comment]
See also: https://stackoverflow.com/questions/33712178/tensorflow-nan-bug/42497444#42497444
